File: examples/quick-git/internal/model/store_test.go
package model

import (
	"context"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
	"time"

	"github.com/egoist/quickgui/go/reactive"
	"quickgui.example/quick-git/internal/git"
)

func refreshTestStore(t *testing.T) (*Store, <-chan func(), func(...string) string, func(string, string)) {
	t.Helper()
	t.Setenv("GIT_CONFIG_GLOBAL", os.DevNull)
	t.Setenv("GIT_CONFIG_NOSYSTEM", "1")
	root := t.TempDir()
	runner := git.NewRunner(2)
	run := func(args ...string) string {
		t.Helper()
		output, err := runner.Text(context.Background(), args, git.CommandOptions{Cwd: root})
		if err != nil {
			t.Fatal(err)
		}
		return strings.TrimSpace(output)
	}
	write := func(name, content string) {
		t.Helper()
		if err := os.WriteFile(filepath.Join(root, name), []byte(content), 0o644); err != nil {
			t.Fatal(err)
		}
	}
	run("init", "-q", "-b", "main")
	run("config", "user.name", "Quick Git Tests")
	run("config", "user.email", "test@example.com")
	run("config", "commit.gpgsign", "false")
	run("config", "core.hooksPath", filepath.Join(root, "no-hooks"))
	write("a.txt", "first\n")
	write("z.txt", "first\n")
	run("add", ".")
	run("commit", "-qm", "Initial")
	write("a.txt", "second\n")
	write("z.txt", "second\n")
	run("commit", "-qam", "Second")
	repo, err := git.Open(context.Background(), runner, root, nil)
	if err != nil {
		t.Fatal(err)
	}
	queue := make(chan func(), 256)
	store := CreateStore(StoreOptions{
		Runner:      runner,
		Persistence: CreatePersistence("", DefaultState(), 0),
		enqueue:     func(fn func()) { queue <- fn },
	})
	t.Cleanup(store.Dispose)
	// Install the fixture without starting native filesystem watchers.
	store.repo, store.main = repo, repo
	reactive.Batch(func() {
		store.setRepository(repo)
		store.setMain(repo)
	})
	return store, queue, run, write
}

func waitStore(t *testing.T, queue <-chan func(), description string, ready func() bool) {
	t.Helper()
	timer := time.NewTimer(5 * time.Second)
	defer timer.Stop()
	for !ready() {
		select {
		case fn := <-queue:
			// Native dispatch applies each turn's writes atomically before effects run.
			reactive.Batch(fn)
		case <-timer.C:
			t.Fatalf("timed out waiting for %s", description)
		}
	}
}

func TestHistoryRefreshRetainsCommitFileAndDiff(t *testing.T) {
	store, queue, run, write := refreshTestStore(t)
	store.SetView(ViewHistory)
	waitStore(t, queue, "initial history diff", func() bool {
		return len(store.History().Commits) == 2 && store.Diff().Diff != nil && !store.Diff().Loading
	})
	store.SelectCommitFile("z.txt")
	waitStore(t, queue, "selected file diff", func() bool {
		diff := store.Diff()
		return diff.Target != nil && diff.Target.Path == "z.txt" && diff.Diff != nil && !diff.Loading
	})
	selection := [][]int{{2, 2}}
	store.SetDiffSelection(selection)
	selectedSHA := store.SelectedCommit().Sha
	parsed := store.Diff().Diff
	detailChanges, diffChanges := 0, 0
	sawHistoryLoading := false
	reactive.CreateRoot(func(dispose func()) struct{} {
		t.Cleanup(dispose)
		reactive.CreateEffect(func() {
			store.CommitDetail()
			detailChanges++
		})
		reactive.CreateEffect(func() {
			store.Diff()
			diffChanges++
		})
		reactive.CreateEffect(func() {
			if store.History().Loading {
				sawHistoryLoading = true
			}
		})
		return struct{}{}
	})
	refresh := func() {
		t.Helper()
		detailChanges, diffChanges = 0, 0
		sawHistoryLoading = false
		generation := store.Generation()
		store.Refresh() // The same path used when the window regains focus.
		waitStore(t, queue, "history refresh", func() bool {
			return store.Generation() > generation && sawHistoryLoading && !store.History().Loading &&
				!store.CommitDetail().Loading && !store.Diff().Loading
		})
		if detailChanges != 0 || diffChanges != 0 {
			t.Fatalf("refresh replaced visible commit details %d times and diff %d times", detailChanges, diffChanges)
		}
		if store.SelectedCommit().Sha != selectedSHA || store.CommitDetail().SelectedPath != "z.txt" ||
			store.Diff().Diff != parsed || !reflect.DeepEqual(store.DiffSelection(), selection) {
			t.Fatal("refresh lost the selected commit, file, diff, or line selection")
		}
	}
	refresh()
	refresh()

	// A commit made while unfocused moves the selected row without changing its contents.
	write("a.txt", "third\n")
	run("commit", "-qam", "Third")
	latestSHA := run("rev-parse", "HEAD")
	refresh()
	if len(store.History().Commits) != 3 || store.History().Commits[0].Sha != latestSHA ||
		!reflect.DeepEqual(store.HistorySelection(), [][]int{{1, 1}}) {
		t.Fatal("refresh did not update history and follow the original commit to its new row")
	}
	store.SelectCommit(0)
	waitStore(t, queue, "newly selected commit", func() bool {
		diff := store.Diff()
		return diff.Target != nil && diff.Target.SHA == latestSHA && diff.Diff != nil && !diff.Loading
	})
	if !strings.Contains(store.Diff().Diff.Render(), "+third") {
		t.Fatal("selecting a different commit retained the previous diff")
	}
}

func TestRefreshUpdatesWorkingTreeDiff(t *testing.T) {
	store, queue, _, write := refreshTestStore(t)
	write("a.txt", "working tree\n")
	store.Refresh()
	waitStore(t, queue, "working tree diff", func() bool {
		return store.Diff().Diff != nil && !store.Diff().Loading
	})
	previous := store.Diff().Diff
	write("a.txt", "changed while unfocused\n")
	store.Refresh()
	waitStore(t, queue, "updated working tree diff", func() bool {
		diff := store.Diff()
		return diff.Diff != nil && diff.Diff != previous && !diff.Loading
	})
	if !strings.Contains(store.Diff().Diff.Render(), "+changed while unfocused") {
		t.Fatal("refresh retained an outdated working tree diff")
	}
}

File: examples/quick-git-typescript/model/store.test.ts
import { afterEach, expect, test } from "bun:test";
import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { createEffect, createRoot, flush } from "solid-js";
import { GitRunner } from "../git/process.ts";
import { createPersistence, DEFAULT_STATE } from "./persistence.ts";
import { createStore, type Store } from "./store.ts";

const cleanup: (() => void | Promise<void>)[] = [];
afterEach(async () => {
  for (const dispose of cleanup.splice(0).reverse()) await dispose();
});
async function until(ready: () => boolean) {
  const deadline = Date.now() + 5000;
  while (Date.now() < deadline) {
    flush();
    if (ready()) return;
    await Bun.sleep(5);
  }
  throw new Error("Model did not settle");
}
async function fixture() {
  const root = await mkdtemp(join(tmpdir(), "quick-git-store-"));
  cleanup.push(() => rm(root, { recursive: true, force: true }));
  const runner = new GitRunner({
    concurrency: 2,
    env: { GIT_CONFIG_GLOBAL: "/dev/null", GIT_CONFIG_NOSYSTEM: "1" },
  });
  const git = (...args: string[]) => runner.text(args, { cwd: root }).then((value) => value.trim());
  const write = (path: string, value: string) => writeFile(join(root, path), value);
  await git("init", "-q", "-b", "main");
  await git("config", "user.name", "Quick Git Tests");
  await git("config", "user.email", "test@example.invalid");
  await git("config", "commit.gpgsign", "false");
  await git("config", "core.hooksPath", "/dev/null");
  await write("a.txt", "first\n");
  await write("z.txt", "first\n");
  await git("add", ".");
  await git("commit", "-qm", "Initial");
  await write("a.txt", "second\n");
  await write("z.txt", "second\n");
  await git("commit", "-qam", "Second");
  const makeStore = () => {
    const store = createStore({
      runner,
      persistence: createPersistence(undefined, DEFAULT_STATE),
      watch: () => () => {},
    });
    cleanup.push(() => store.dispose());
    return store;
  };
  const store = makeStore();
  expect(await store.openRepository(root)).toBe(true);
  flush();
  return { store, runner, git, write, root, makeStore };
}
function diffReady(store: Store) {
  return store.diff().files.length > 0 && !store.diff().loading;
}
function diffText(store: Store) {
  return store
    .diff()
    .files.flatMap((file) => file.hunks.flatMap((hunk) => hunk.lines.map((line) => line.text)))
    .join("\n");
}

test("focus refresh retains the commit, file, diff identity, and selected lines", async () => {
  const { store, git, write } = await fixture();
  store.setView("history");
  await until(() => store.history().commits.length === 2 && diffReady(store));
  store.selectCommitFile("z.txt");
  await until(() => store.diff().target?.path === "z.txt" && diffReady(store));
  store.setDiffSelection([[2, 2]]);
  flush();
  const selectedSHA = store.selectedCommit()!.sha;
  const detail = store.commitDetail();
  const diff = store.diff();
  let detailChanges = 0;
  let diffChanges = 0;
  cleanup.push(
    createRoot((dispose) => {
      createEffect(store.commitDetail, () => {
        detailChanges++;
      });
      createEffect(store.diff, () => {
        diffChanges++;
      });
      return dispose;
    }),
  );
  flush();
  async function refresh() {
    detailChanges = 0;
    diffChanges = 0;
    await store.refresh();
    await until(() => !store.history().loading && diffReady(store));
    expect(store.selectedCommit()!.sha).toBe(selectedSHA);
    expect(store.commitDetail()).toBe(detail);
    expect(store.diff()).toBe(diff);
    expect(store.diffSelection()).toEqual([[2, 2]]);
    expect({ detailChanges, diffChanges }).toEqual({ detailChanges: 0, diffChanges: 0 });
  }
  await refresh();
  await refresh();
  await write("a.txt", "third\n");
  await git("commit", "-qam", "Third");
  await refresh();
  expect(store.history().commits).toHaveLength(3);
  expect(store.historySelection()).toEqual([[1, 1]]);
  store.setHistorySelection([[0, 0]]);
  await until(
    () => store.diff().target?.sha === store.history().commits[0]!.sha && diffReady(store),
  );
  expect(diffText(store)).toContain("third");
});

test("working-tree refresh replaces stale content and preserves paths across reordered rows", async () => {
  const { store, write } = await fixture();
  await write("z.txt", "working tree\n");
  await store.refresh();
  await until(() => store.diff().target?.path === "z.txt" && diffReady(store));
  store.setSelection("unstaged", [[0, 0]]);
  flush();
  const previous = store.diff().files;
  await write("a.txt", "another change\n");
  await write("z.txt", "changed while unfocused\n");
  await store.refresh();
  await until(() => store.diff().files !== previous && !store.diff().loading);
  expect(store.selectedItems("unstaged").map((item) => item.path)).toEqual(["z.txt"]);
  expect(store.diff().target?.path).toBe("z.txt");
  expect(diffText(store)).toContain("changed while unfocused");
});

test("all-branches changes apply to the immediate history request", async () => {
  const { store, git, write } = await fixture();
  await git("switch", "-qc", "feature");
  await write("feature.txt", "feature\n");
  await git("add", ".");
  await git("commit", "-qm", "Feature only");
  await git("switch", "-q", "main");
  await store.refresh();
  store.setHistoryAllBranches(true);
  await until(() => !store.history().loading && store.history().commits.length === 3);
  expect(store.history().commits.some((commit) => commit.subject === "Feature only")).toBe(true);
  store.setHistoryAllBranches(false);
  await until(() => !store.history().loading && store.history().commits.length === 2);
});

test("a successful commit clears the composer and refreshes the repository", async () => {
  const { store, write } = await fixture();
  await write("a.txt", "third\n");
  await store.refresh();
  await store.stageAll();
  store.setSubject("Third");
  store.setBody("New body");
  flush();
  expect(store.canCommit()).toBe(true);
  await store.commit();
  flush();
  expect(store.subject()).toBe("");
  expect(store.body()).toBe("");
  expect(store.staged()).toHaveLength(0);
  expect(store.history().commits[0]!.subject).toBe("Third");
});

test("closing during repository loading cancels jobs and cannot repopulate a closed store", async () => {
  const { store, runner, root } = await fixture();
  const opened = store.openRepository(root);
  store.closeRepository();
  expect(await opened).toBe(false);
  await until(() => runner.pending === 0);
  expect(store.repository()).toBeUndefined();
  expect(store.commitDetail().files).toEqual([]);
  const pending = store.openRepository(root);
  store.dispose();
  expect(await pending).toBe(false);
  await until(() => runner.pending === 0);
});

test("repository stores opened from another root have independent lifetimes", async () => {
  const { root, makeStore } = await fixture();
  let other!: Store;
  const disposeSource = createRoot((dispose) => {
    other = makeStore();
    return dispose;
  });
  expect(await other.openRepository(root)).toBe(true);
  disposeSource();
  other.setView("history");
  await until(() => diffReady(other));
  expect(other.history().commits).toHaveLength(2);
});
